daemon/src/engine/server/CryptoChallege.h

```diff
@@ -0,0 +1,633 @@
+    {
+        return valid_until >= time;
+    }
+
+    /*
+     * Whether two challenges match
+     */
+    bool Matches( const Challenge& other ) const
+    {
+        if ( challenge != other.challenge )
+            return false;
+
+        return NET_CompareAdr( source, other.source );
+    }
+
+    /*
+     * Challenge as a hex string
+     */
+    std::string String() const
+    {
+        return Crypto::Encoding::HexEncode( challenge );
+    }
+
+private:
+    /*
+     * Generates a random challenge
+     */
+    std::string GenerateString()
+    {
+        std::vector<uint8_t> data( Bytes() );
+        Sys::GenRandomBytes(data.data(), data.size());
+        std::ostringstream stream;
+        stream.setf(std::ios::hex, std::ios::basefield);
+        stream.fill('0');
+        for ( auto ch : data )
+        {
+            stream.width(2);
+            stream << int( ch );
+        }
+        return stream.str();
+    }
+
+    TimePoint valid_until;
+    Crypto::Data challenge;
+    netadr_t source;
+
+};
+
+class ChallengeManager
+{
+public:
+    static ChallengeManager& Instance()
+    {
+        static ChallengeManager singleton;
+        return singleton;
+    }
+
+    int MaxChallenges() const
+    {
+        // TODO Read this from a cvar
+        return 1024;
+    }
+
+    /*
+     * Generates a challenge for the given address and returns the challenge string
+     */
+    std::string GenerateChallenge( const netadr_t& source )
+    {
+        auto challenge = Challenge( source );
+        Push( challenge );
+        return challenge.String();
+    }
+
+    /*
+     * Add a challenge to the pool
+     */
+    void Push( const Challenge& challenge )
+    {
+        auto lock = Lock();
+
+        Cleanup();
+
+        if ( challenges.size() >= MaxChallenges() )
+        {
+            challenges.pop_front();
+        }
+
+        challenges.push_back( challenge );
+    }
+
+    /*
+     * Check if a challenge matches any of the registered ones
+     * If it does and it's valid, it'll return false
+     */
+    bool Match( const Challenge& challenge )
+    {
+        auto lock = Lock();
+
+        Cleanup();
+
+        auto it = std::find_if( challenges.begin(), challenges.end(),
+            [&challenge]( const Challenge& ch ) {
+                return ch.Matches( challenge );
+        } );
+
+        if ( it != challenges.end() )
+        {
+            challenges.erase( it );
+            return true;
+        }
+
+        return false;
+    }
+
+private:
+    ChallengeManager() = default;
+    ChallengeManager( const ChallengeManager& ) = delete;
+    ChallengeManager& operator=( const ChallengeManager& ) = delete;
+
+    std::unique_lock<std::mutex> Lock()
+    {
+        return std::unique_lock<std::mutex>{mutex};
+    }
+
+    /*
+     * Removes outdated challenges
+     * PRE: The caller has acquired a lock on mutex
+     */
+    void Cleanup()
+    {
+        auto now = Challenge::Clock::now();
+        challenges.erase(
```

Comment:
fails if the deque is empty, that's a strong caveat of the remove-erase_if idiom.